File: single_thread/dart/prime.dart
class PrimeFinder {
  bool isPrime(int number) {
    if (number <= 1) return false;
    if (number <= 3) return true;
    
    if (number % 2 == 0) return false;
    
    for (int i = 3; i * i <= number; i += 2) {
      if (number % i == 0) return false;
    }
    
    return true;
  }

  List<int> findPrimesInRange(int start, int end) {
    List<int> primes = [];
    
    for (int number = max(start, 2); number <= end; number++) {
      if (isPrime(number)) {
        primes.add(number);
      }
    }
    
    return primes;
  }

  // Helper method to find maximum of two numbers
  int max(int a, int b) {
    return a > b ? a : b;
  }
}

// Example usage
void main() {
  final primeFinder = PrimeFinder();
  
  final stopwatch = Stopwatch()..start();
  final primes = primeFinder.findPrimesInRange(1, 20000000);
  print('Primes count: ${primes.length}');
  print('Time elapsed: ${stopwatch.elapsed} (ms)');
}

File: single_thread/go/prime.go
package main

import (
	"fmt"
	"math"
	"time"
)

// isPrime checks if a given number is prime
func isPrime(n int) bool {
	// Numbers less than 2 are not prime
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n % 2 == 0 {
		return false
	}
	
	// Check for divisibility up to the square root of the number
	// This optimization reduces unnecessary iterations
	sqrtN := int(math.Sqrt(float64(n)))
	
	for i := 3; i <= sqrtN; i+=2 {
		if n % i == 0 {
			return false
		}
	}
	
	return true
}

// findPrimesInRange finds all prime numbers within a specified range
func findPrimesInRange(start, end int) []int {
	primes := []int{}
	
	for num := start; num <= end; num++ {
		if isPrime(num) {
			primes = append(primes, num)
		}
	}
	
	return primes
}

func main() {
	startTime := time.Now()

	// Example usage: find primes between 1 and 50
	start, end := 1, 20000000
	primesFound := findPrimesInRange(start, end)
	duration := time.Since(startTime)
	
	fmt.Printf("Prime numbers between %d and %d:\n", start, end)

	fmt.Printf("Total primes found: %d\n", len(primesFound))
	fmt.Printf("Time taken: %v\n", duration)

	fmt.Println()
}